### DISCUSSION TITLE
[11.0.0] ItemsControl duplicating elements after clearing the value bound to ItemsSource and adding them back

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#12137] [Q&A] [answered]
I'm getting a strange issue in my app where elements are getting duplicated **almost** each and every time i clear, or re-instantiate the `ObervableCollection`, in my case, it happens every time my Application successfully connect to another, specific application.

https://github.com/AvaloniaUI/Avalonia/assets/39861216/43c6c323-f42f-4b02-9372-70e808b2774d

Here, in the right column of my application, you can see the Advanced Mode section, it contains multiple elements, one of them being the `ItemsControl` in question.
I also have on my ViewModel a property call `Display`, which is an `ObervableCollection<WheelBindingDisplayViewModel>`.
This Collection is cleared every time i build my interface as you can see in the recording.

The process of re-creating this should be farily simple: 

- Add your initial elements through whatever means,
- Clear the collection,
- Add them back,
- Duplicated ContentPresenters.

I assume this is a bug in avalonia considering many of the issues that appeared with the very early 11.0.0 release,
plus i'm not doing anything out of the ordinary so it's definitely not my fault.

Any idea on whether it's an actual issue and how i could workaround it?
Thanks in advance.

EDIT: Reproducible example: https://github.com/Mrcubix/Avalonia-ItemsControl-Issue

<p align="center">
    <img src="https://github.com/AvaloniaUI/Avalonia/assets/39861216/a06111de-d645-4219-a639-636f4225f797" alt="Avalonia bug duplicating ContentPresenter elements" />
</p>

**Comment by Mrcubix** [comment]
![image](https://github.com/AvaloniaUI/Avalonia/assets/39861216/491d8fed-c097-44dc-88c9-275d563cca68)

![image](https://github.com/AvaloniaUI/Avalonia/assets/39861216/161d329d-b182-4f8f-bef3-87ad02ce155e)

One interesting thing is that the `Capacity` of the `VisualChildren` of `ItemsControl` indeed reflect this issue, although `ItemCount` is 2

One more thing to note is that the parent of this ItemsControl is a `ScrollViewer`, and i have seen a couple issues around that element here.

**Reply by Mrcubix** [reply]
this seems more likely to be an Avalonia bug for this specific control as `ItemsRepeater` works without any issues on that side.

**Reply by Mrcubix** [reply]
was able to replicate it in a easily reproducible example repo, will send it soon:

<p align="center">
    <img src="https://github.com/AvaloniaUI/Avalonia/assets/39861216/a06111de-d645-4219-a639-636f4225f797" alt="Avalonia bug duplicating ContentPresenter elements" />
</p>

it seem to be caused when clearing and adding elements within another thread

**Comment by timunie** [comment]
Any chance you can file a minimal repo? I cannot manage to reproduce it in a simple sample on my end.

**Reply by Mrcubix** [reply]
I'll try cooking something simillar to my use case and come back to you, though it seems like an actual avalonia bug as ItemsRepeater works properly, i have seen it being suggested around here instead of ItemsControl.

**Reply by Mrcubix** [reply]
Here you go: https://github.com/Mrcubix/Avalonia-ItemsControl-Issue

**Reply by timunie** [reply, marked as answer]
updating a collection from another thread is not allowed iirc. Use `Disptcher.Post()` in order to update the collection. The rest can be done on background thread. You can also use DynamicData which allows it out of the box by `ObserveOn(RxApp.MainThead)`.

**Reply by timunie** [reply]
https://github.com/Mrcubix/Avalonia-ItemsControl-Issue/pull/1

**Reply by Mrcubix** [reply]
> updating a collection from another thread is not allowed iirc. Use `Disptcher.Post()` in order to update the collection. The rest can be done on background thread. You can also use DynamicData which allows it out of the box by `ObserveOn(RxApp.MainThead)`.

then an exception should probably be thrown, like Avalonia does when trying to open a dialog outside of the UI thread.
will have to ask explanation on what some of the code you used does, but that will be done on the PR.
